File: haskell/progetto_batch.hs
{- Importazione delle librerie -}
import Text.Printf (printf)
import Data.List (sortOn)
import Data.Function (on)
import System.IO (readFile)

{- Definizione di un tipo per rappresentare punti in uno spazio 2D -}
data Point2D = Point2D { xCoord :: Double, yCoord :: Double } deriving Show

{- Definizione di un tipo per rappresentare i punti 2D etichettati con una classe (di tipo Int) -}
data PointLabeled = PointLabeled { x :: Double, y :: Double, label :: Int }

{- CALCOLO DELLA REGRESSIONE LINEARE -}

{- Calcola la media di una lista di Double -}
mean :: [Double] -> Double
mean xs = sum xs / fromIntegral (length xs)

{- Calcola la varianza di una lista di Double -}
variance :: [Double] -> Double
variance xs = sum [(x - m) ** 2 | x <- xs] / fromIntegral (length xs)
  where
    m = mean xs

{- Calcola la covarianza di due liste di Double -}
covariance :: [Double] -> [Double] -> Double
covariance xs ys = sum [(x - mx) * (y - my) | (x, y) <- zip xs ys] / fromIntegral (length xs)
  where
    mx = mean xs
    my = mean ys

{- Calcola i coefficienti (pendenza e intercetta) della retta interpolatrice -}
linearRegression :: [Point2D] -> (Double, Double)
linearRegression points = (slope, intercept)
  where
    xs = map xCoord points
    ys = map yCoord points
    slope = covariance xs ys / variance xs
    intercept = mean ys - slope * mean xs

{- Funzione per ottenere dall'utente una lista di punti separati da spazio -}
readPoints :: IO [Point2D]
readPoints = do
  putStrLn "Enter the dataset in the format 'x1 y1 x2 y2 ... xn yn':"
  input <- getLine
  let coordinates = map read $ words input
  let points = makePoints coordinates
  if length points < 2
    then do
      putStrLn "At least two points are required for linear regression."
      readPoints
    else return points

{- Funzione che converte una lista di Double in una lista di Point2D -}
makePoints :: [Double] -> [Point2D]
makePoints [] = []
makePoints (x:y:rest) = Point2D x y : makePoints rest
makePoints _ = error "Invalid number of coordinates."

{- Funzione che ottiene dall'utente il valore di 'x' da valutare sulla retta ottenuta -}
readXToEvaluate :: IO (Maybe Double)
readXToEvaluate = do
  putStrLn "Enter the value of 'x' to evaluate the fitted line (or 'q' to quit):"
  input <- getLine
  if input == "q"
    then return Nothing
    else do
      let x = read input
      return $ Just x

{- Funzione che valuta la regressione lineare per il punto -}
evaluatePoint :: Point2D -> (Double, Double) -> Double
evaluatePoint (Point2D x _) (slope, intercept) = slope * x + intercept

{- CALCOLO DEL KNN -}

{- Funzione che calcola la distanza euclidea tra due punti -}
distance :: PointLabeled -> PointLabeled -> Double
distance p1 p2 = sqrt $ (x p1 - x p2) ^ 2 + (y p1 - y p2) ^ 2

{- Funziona che trova i k vicini per un punto dal dataset -}
kNearestNeighbors :: Int -> PointLabeled -> [PointLabeled] -> [PointLabeled]
kNearestNeighbors k testPoint dataset =
  take k $ sortOn (distance testPoint) dataset

{- Funzione che trova la classe di maggioranza tra i vicini -}
majorityClass :: [PointLabeled] -> Int
majorityClass neighbors =
  let counts = map (\label' -> (label', length (filter (\p -> label p == label') neighbors))) [0, 1, 2]
      sortedCounts = reverse $ sortOn snd counts
  in fst $ head sortedCounts

{- Funzione per ottenere dall'utente un dataset di punti etichettati con una classe -}
readLabeledDataset :: Int -> [PointLabeled] -> IO [PointLabeled]
readLabeledDataset 0 acc = return $ reverse acc
readLabeledDataset n acc = do
  input <- getLine
  let [x, y, label] = map read $ words input
  readLabeledDataset (n-1) (PointLabeled x y (round label) : acc)  -- Convert label to Int using `round`

{- Funzione per ottenere dall'utente un punto da testare sul KNN -}
readPoint :: IO PointLabeled
readPoint = do
  input <- getLine
  let [x, y] = map read $ words input
  return $ PointLabeled x y 0  -- The class label for the test point is not relevant, so we set it to 0.

{- Function to read points from the hardcoded file for linear regression -}
readPointsFromFile :: IO [Point2D]
readPointsFromFile = do
  contents <- readFile "linear_regression_dataset.txt"
  let coordinates = map read $ words contents
  let points = makePoints coordinates
  if length points < 2
    then do
      putStrLn "At least two points are required for linear regression."
      readPoints
    else return points

{- Function to read labeled dataset from the hardcoded file for k-Nearest Neighbors (KNN) -}
readLabeledDatasetFromFile :: IO [PointLabeled]
readLabeledDatasetFromFile = do
  contents <- readFile "knn_dataset.txt"
  let entries = map words $ lines contents
  return $ map (\[x, y, label] -> PointLabeled (read x) (read y) (read label)) entries

{- Helper function to print a horizontal line -}
printHorizontalLine :: IO ()
printHorizontalLine = putStrLn "---------------------------------------------------------------"

{- Funzione Main contenente il menu di selezione dell'operazione da svolgere -}
mainMenu :: IO ()
mainMenu = do
  putStrLn ""
  putStrLn "Choose the operation to compute:"
  putStrLn "1 - Linear Regression"
  putStrLn "2 - k-Nearest Neighbors (KNN)"
  putStrLn "3 - Quit"

  choice <- getLine
  case choice of
    "1" -> do
      putStrLn ""
      putStrLn "-------- Linear Regression --------"
      putStrLn ""
      putStrLn "Fitting the best line to the dataset..."
      putStrLn ""      
      
      points <- readPointsFromFile
      let (slope, intercept) = linearRegression points
      let formattedSlope = printf "%.2f" slope
      let formattedIntercept = printf "%.2f" intercept
      
      putStrLn $ "Best-fitting line: y = " ++ formattedSlope ++ "x + " ++ formattedIntercept
      putStrLn ""

      -- Ask the user for the value of 'x' to evaluate the fitted line
      testX <- readXToEvaluate
      case testX of
        Nothing -> putStrLn "Exiting the program."
        Just x -> do
          let predictedY = evaluatePoint (Point2D x 0) (slope, intercept)
          
          putStrLn ""
          putStrLn $ "Predicted y for x=" ++ show x ++ " is " ++ show predictedY
          putStrLn ""
          printHorizontalLine
          
          mainMenu

    "2" -> do
      putStrLn "----- k-Nearest Neighbors (KNN) -----"
      putStrLn "Enter the value of k:"
      k <- readLn :: IO Int

      putStrLn "Enter the test point (format: x y):"
      testPoint <- readPoint

      dataset <- readLabeledDatasetFromFile
      let kNearest = kNearestNeighbors k testPoint dataset
          testClass = majorityClass kNearest

      putStrLn $ "Predicted class for the test point: " ++ show testClass
      mainMenu
    "3" -> putStrLn "Exiting the program."
    _ -> do
      putStrLn "Invalid choice. Please select a valid option."
      mainMenu

{- Punto iniziale del programma -}
main :: IO ()
main = do
  printHorizontalLine
  putStrLn "Progetto del corso di Programmazione Logica e Funzionale"
  putStrLn "Anno 2022/2023"
  putStrLn "Progetto realizzato da: Andrea De Lorenzis"
  printHorizontalLine
  
  mainMenu

File: prolog/progetto_batch.pl
% ESECUZIONE MAIN

/* Punto iniziale del programma */
main :-
  nl, nl, 
  write('Progetto del corso di Programmazione Logica e Funzionale'), nl,
  write('Anno 2022/2023'), nl,
  write('Progetto realizzato da: Andrea De Lorenzis'), nl, nl, nl,
  main_menu.
  
/* Menu di selezione dell'operazione da svolgere */
main_menu :-
  write('Selezionare l\'operazione da svolgere: '), nl,
  write('1 - Linear Regression'), nl,
  write('2 - K-Nearest Neighbors (KNN)'), nl,
  write('3 - Quit'), nl, nl,
  
  read(Choice),
  (Choice = 1 -> doLinearRegression, main_menu ;
   Choice = 2 -> doKNN, main_menu ;
   Choice = 3 -> write('Chiusura del programma.'), nl ;
   write('Scelta incorretta. Per favore, seleziona una delle opzioni valide.'), nl, 
   main_menu). 
   
/* Funzione per eseguire la regressione lineare */
doLinearRegression :-
  readDataset(Points),
  nl, nl, write(Points), nl, nl.
  
  /*linearRegression(Points, Slope, Intercept), 
  nl,
  format('Best-fitting line: y = ~2f*x + ~2f~n', [Slope, Intercept]), nl,
  evaluateXValues(Slope, Intercept).*/
 
% CALCOLO REGRESSIONE LINEARE

/* Calcola la media di una lista di numeri */
mean([], 0).
mean([X | Xs], Mean) :-
    mean(Xs, RestMean),
    length(Xs, N),
    Mean is (X + N * RestMean) / (N + 1).
  
/* Calcola la varianza di una lista di numeri */  
variance([], 0).
variance([X | Xs], Var) :-
    mean([X | Xs], Mean),
    variance(Xs, RestVar),
    Var is RestVar + (X - Mean) * (X - Mean).
	
/* Calcola la covarianza di due liste di numeri */
covariance([], [], 0).
covariance([X | Xs], [Y | Ys], Cov) :-
  mean([X | Xs], MeanX),
  mean([Y | Ys], MeanY),
  covariance(Xs, Ys, RestCov),
  Cov is RestCov + (X - MeanX) * (Y - MeanY).
  
/* Calcola i coefficienti (pendenza e intercetta) della retta interpolatrice */
linearRegression(Points, Slope, Intercept) :-
  extractXCoords(Points, Xs),
  extractYCoords(Points, Ys),
  variance(Xs, VarX),
  covariance(Xs, Ys, CovXY),
  Slope is CovXY / VarX,
  mean(Ys, MeanY),
  mean(Xs, MeanX),
  Intercept is MeanY - Slope * MeanX.
 
/* Estra le coordinate X da una lista di punti */
extractXCoords([], []).
extractXCoords([(X, _) | Points], [X | Xs]) :-
  extractXCoords(Points, Xs).
  
/* Estrae le coordinate Y da una lista di punti */
extractYCoords([], []).
extractYCoords([(_, Y) | Points], [Y | Ys]) :-
  extractYCoords(Points, Ys).
  
/* Predicato per leggere una lista di tuple da file */
readDataset(List) :-
  open('dataset.txt', read, Stream),
  read_line(Stream, Tuple),
  close(Stream).

read_digit(Stream, Digit) :-
    read(Stream, Digit), integer(Digit), Digit >= 1, Digit =< 9.

read_line(end_of_file, _) :- !.
  
read_line(Stream, Tuple) :-
  read_digit(Stream, X),
  read_digit(Stream, Y),
  write(X), write(' '), write(Y),
  Tuple is (X, Y).

process_stream(end_of_file, _) :- !.
 
process_stream(Char, Stream) :-
        write(Char),
        get_char(Stream, Char2),
        process_stream(Char2, Stream).

/*
  nl, write('Inserisci i punti del dataset nel formato: [(x1,y1), ..., (xn,yn)].'), nl, nl,
  read(List),
  length(List, NumPoints),
  (NumPoints < 2 -> write('Sono necessari almeno due punti per la regressione lineare'), readDataset([]), nl ; true).
  */
  
/* Funzione che ottiene dall'utente una coordinata X da valutare sulla retta ottenuta */
readXValue(Value) :-
  write('Inserisci un valore per la coordinata x (o \'q\' per tornare al menu): '), nl, nl,
  read(Value).
  
/* Predicato che cicla continuamente per valutare i punti fin a quando 
   l'utente non inserisce 'q' */
evaluateXValues(Slope, Intercept) :-
  readXValue(Value),
  (Value = 'q' -> true ;
   evaluateXValue(Value, Slope, Intercept, Y),
   nl, format('Il valore di Y previsto e\': ~2f', [Y]), nl,
   evaluateXValues(Slope, Intercept)).

/* Funzione che valuta la regressione lineare per il punto */
evaluateXValue(X, Slope, Intercept, Y) :-
  Y is Slope * X + Intercept.

% CALCOLO K-NEAREST NEIGHBORS

/* Funzione che calcola la distanza euclidea tra due punti */
distance((X1, Y1, _), (X2, Y2), Dist) :-
  Dist is sqrt((X1 - X2) * (X1 - X2) + (Y1 - Y2) * (Y1 - Y2)).
  
/* Funzione che trova i k vicini per un punto */
kNearestNeighbors(K, TestPoint, Dataset, Neighbors) :-
  calculateDistances(TestPoint, Dataset, Distances),
  sortDistances(Distances, SortedDistances),
  take(K, SortedDistances, NeighborsPairs),
  convertToNeighbors(NeighborsPairs, Neighbors).
  
/* Rimuove il valore della distanza dal formato (Dist-(X, Y, Class)) 
   per ottenere il formato (X, Y, Class) */
convertToNeighbors([], []).
convertToNeighbors([_-Neighbor | RestPairs], [Neighbor | RestNeighbors]) :-
    convertToNeighbors(RestPairs, RestNeighbors).

/* Predicato per calolcare le distanze tra TestPoint e ogni altro punto del dataset */
calculateDistances(_, [], []).
calculateDistances(TestPoint, [Point | Rest], [Dist-Point | Distances]) :-
  distance(Point, TestPoint, Dist),
  calculateDistances(TestPoint, Rest, Distances).
  
/* Ordina le distanze in ordine crescente */
sortDistances(Distances, SortedDistances) :-
  keysort(Distances, SortedDistances).

/* Prende i primi K elementi di una lista */
take(0, _, []).
take(K, [X | Xs], [X | Rest]) :-
  K > 0,
  K1 is K - 1,
  take(K1, Xs, Rest).
  
/* Funzione che trova la classe di maggioranza tra i vicini */
majorityClass(Neighbors, MajorityClass) :-
    countClassOccurrences(Neighbors, CountMap),
    maxClassOccurrences(CountMap, MajorityClass).

/* Predicato ausiliario per conteggiare le occorrenze di ogni 
  classe nei vicini */
countClassOccurrences(Neighbors, CountMap) :-
    countClassOccurrences(Neighbors, [], CountMap).

countClassOccurrences([], CountMap, CountMap).
countClassOccurrences([(X, Y, Class) | Neighbors], PartialCountMap, CountMap) :-
    updateCountMap(Class, PartialCountMap, NewCountMap),
    countClassOccurrences(Neighbors, NewCountMap, CountMap).

updateCountMap(Class, [], [(Class, 1)]).
updateCountMap(Class, [(Class, Count) | Rest], [(Class, NewCount) | Rest]) :-
    NewCount is Count + 1.
updateCountMap(Class, [(OtherClass, Count) | Rest], [(OtherClass, Count) | NewRest]) :-
    Class \= OtherClass,
    updateCountMap(Class, Rest, NewRest).

/* Predicato ausiliario per trovare la classe con il maggior numero di 
  occorrenze */
maxClassOccurrences([(Class, Count) | Rest], MajorityClass) :-
    maxClassOccurrences(Rest, Class, Count, MajorityClass).

maxClassOccurrences([], MajorityClass, _, MajorityClass).
maxClassOccurrences([(Class, Count) | Rest], MaxClass, MaxCount, MajorityClass) :-
    ( Count > MaxCount ->
        maxClassOccurrences(Rest, Class, Count, MajorityClass)
    ;   maxClassOccurrences(Rest, MaxClass, MaxCount, MajorityClass)
    ).

/* Funzione per ottenere dall'utente un dataset di punti etichettati con 
   una classe */
readLabeledDataset(List) :-
  nl, write('Inserisci i punti del dataset nel formato: [(x1,y1,C), ..., (xn,yn,C)].'), nl, nl,
  read(List), nl,
  length(List, NumPoints),
  (NumPoints < 2 -> write('Sono necessari almeno due punti per la regressione lineare'), readLabeledDataset([]), nl ; true).
  
/* Funzione per leggere un singolo punto da valutare con KNN */
readLabeledPoint(Point) :-
  write('Inserisci il valore del punto da testare nel formato (x,y): '), nl, nl,
  read(Point).
  
/* Funzione per eseguire il KNN */
doKNN :- 
  readLabeledDataset(Dataset),
  nl, write('Inserisci il valore per k: '), nl, nl,
  read(K), nl, nl,
  evaluatePoints(Dataset, K).

/* Funzione per continuamente inserire punti da valutare con KNN */
evaluatePoints(Dataset, K) :-
  readLabeledPoint(TestPoint),
  kNearestNeighbors(K, TestPoint, Dataset, Neighbors),
  displayNeighbors(Neighbors),
  majorityClass(Neighbors, Class),
  intToChar(ClassInt, Class),
  nl, nl, format('La classe prevista per il punto è: ~w~n', [Class]), 
  nl, nl, write('Vuoi continuare? (si/no)'), nl, nl,
  read(Choice),
  (Choice = 'si' -> evaluatePoints(Dataset, K) ; true).
  
displayNeighbors([]).
displayNeighbors([Dist-Point | Rest]) :-
  Point = (X, Y, ClassInt),
  intToChar(ClassInt, Class),
  format('Distanza: ~2f, Punto: (~2f, ~2f, ~w)~n', [Dist, X, Y, Class]),
  displayNeighbors(Rest).
  
% Predicate to convert integer class values to characters
intToChar(0, 'A').
intToChar(1, 'B').
intToChar(2, 'C').
intToChar(3, 'D').
